### DISCUSSION TITLE
Docs: `output: "standalone"` conflicts with discouraged usage of Custom Servers

### DISCUSSION THREAD

**Discussion by larsqa** [vercel/next.js#42603] [Help]
### What is the improvement or update you wish to see?

The [Custom Server documentation](https://nextjs.org/docs/advanced-features/custom-server) highlights

> A custom server will remove important performance optimizations, like serverless functions and [Automatic Static Optimization](https://nextjs.org/docs/advanced-features/automatic-static-optimization).

However, [`output:"standalone"`](https://nextjs.org/docs/advanced-features/output-file-tracing#automatically-copying-traced-files) builds a custom server inside `.next/standalone/server.js` which is used in the [official Docker examples](https://github.com/vercel/next.js/blob/canary/examples/with-docker/next.config.js#L3).

I'm not sure if this is intended, a bug, or requires additional documentation, but this seems very confusing.

Why would a Next.js feature yield something that is a Next.js "antipattern"?

**Comment by michal-mad** [comment]
It's not that 'custom' as according to the documentation:

> Additionally, a minimal server.js file is also output which can be used instead of next start. [Reference](https://nextjs.org/docs/advanced-features/output-file-tracing#automatically-copying-traced-files)

And in fact it does look like minimal version of nextjs server if you look inside `server.js`. I am assuming it does provide all nice features just like it is with 'normal' `next start`.

**Comment by larsqa** [comment]
Here's the minimal server.js which Next.js currently (12.2.2) outputs:
```js

process.env.NODE_ENV = 'production'
process.chdir(__dirname)
const NextServer = require('next/dist/server/next-server').default
const http = require('http')
const path = require('path')

// Make sure commands gracefully respect termination signals (e.g. from Docker)
// Allow the graceful termination to be manually configurable
if (!process.env.NEXT_MANUAL_SIG_HANDLE) {
  process.on('SIGTERM', () => process.exit(0))
  process.on('SIGINT', () => process.exit(0))
}

let handler

const server = http.createServer(async (req, res) => {
  try {
    await handler(req, res)
  } catch (err) {
    console.error(err);
    res.statusCode = 500
    res.end('internal server error')
  }
})
const currentPort = parseInt(process.env.PORT, 10) || 3000

server.listen(currentPort, (err) => {
  if (err) {
    console.error("Failed to start server", err)
    process.exit(1)
  }
  const nextServer = new NextServer({
    hostname: 'localhost',
    port: currentPort,
    dir: path.join(__dirname),
    dev: false,
    conf: {"env":{},"webpackDevMiddleware":null,"eslint":{"ignoreDuringBuilds":false},"typescript":{"ignoreBuildErrors":false,"tsconfigPath":"tsconfig.json"},"distDir":"./.next","cleanDistDir":true,"assetPrefix":"","configOrigin":"next.config.js","useFileSystemPublicRoutes":true,"generateEtags":true,"pageExtensions":["tsx","ts","jsx","js"],"target":"server","poweredByHeader":true,"compress":true,"analyticsId":"","images":{"deviceSizes":[480,768,992,1280,1536],"imageSizes":[16,32,48,64,96,128,256,384],"path":"/_next/image","loader":"default","domains":["images.ctfassets.net"],"disableStaticImages":false,"minimumCacheTTL":60,"formats":["image/webp"],"dangerouslyAllowSVG":false,"contentSecurityPolicy":"script-src 'none'; frame-src 'none'; sandbox;"},"devIndicators":{"buildActivity":true,"buildActivityPosition":"bottom-right"},"onDemandEntries":{"maxInactiveAge":15000,"pagesBufferLength":2},"amp":{"canonicalBase":""},"basePath":"","sassOptions":{},"trailingSlash":false,"i18n":null,"productionBrowserSourceMaps":false,"optimizeFonts":true,"excludeDefaultMomentLocales":true,"serverRuntimeConfig":{},"publicRuntimeConfig":{},"reactStrictMode":true,"httpAgentOptions":{"keepAlive":true},"outputFileTracing":true,"staticPageGenerationTimeout":60,"swcMinify":false,"output":"standalone","experimental":{"manualClientBasePath":false,"legacyBrowsers":true,"browsersListForSwc":false,"newNextLinkBehavior":false,"cpus":9,"sharedPool":true,"profiling":false,"isrFlushToDisk":true,"workerThreads":false,"pageEnv":false,"optimizeCss":false,"nextScriptWorkers":false,"scrollRestoration":false,"externalDir":false,"disableOptimizedLoading":false,"gzipSize":true,"swcFileReading":true,"craCompat":false,"esmExternals":true,"appDir":false,"isrMemoryCacheSize":52428800,"serverComponents":false,"fullySpecified":false,"outputFileTracingRoot":"","images":{"remotePatterns":[]},"swcTraceProfiling":false,"forceSwcTransforms":false,"largePageDataBytes":128000,"trustHostHeader":false},"configFileName":"next.config.js"},
  })
  handler = nextServer.getRequestHandler()

  console.log("Listening on port", currentPort)
})
    
```
Above standalone server.js has similarities with:
- [what's executed during `next start`](https://github.com/vercel/next.js/blob/canary/packages/next/cli/next-start.ts), which [eventually uses the similar looking `start-server.ts`](https://github.com/vercel/next.js/blob/canary/packages/next/server/lib/start-server.ts).
- the [custom server example](https://nextjs.org/docs/advanced-features/custom-server)

However, what's the key element that removes/add's the Next.js features lost with a custom server:

> Before deciding to use a custom server, please keep in mind that it should only be used when the integrated router of Next.js can't meet your app requirements. A custom server will remove important performance optimizations, like serverless functions and [Automatic Static Optimization](https://nextjs.org/docs/advanced-features/automatic-static-optimization).
>
> [Source](https://nextjs.org/docs/advanced-features/custom-server)

**Comment by ijjk** [comment]
Hi, to clarify here, the minimal server file output for `output: 'standalone'` is not a custom server as it is still managed by Next.js and doesn't leverage any features that aren't supported directly by `next start`.

**Comment by larsqa** [comment]
@ijjk, thanks for the clarification! 

Am I then right to understand that we can use the minimal server from the standalone output in a project that requires a custom server? This would then prevent losing *important performance optimizations, like serverless functions and Automatic Static Optimization* as outlined in [custom server](https://nextjs.org/docs/advanced-features/custom-server)?
